Code/Framework/AzCore/AzCore/std/containers/span.h

```diff
@@ -0,0 +1,159 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+#pragma once
+
+#include <AzCore/std/containers/vector.h>
+#include <AzCore/std/containers/fixed_vector.h>
+#include <AzCore/std/containers/array.h>
+
+namespace AZStd
+{
+    /**
+     * First pass partial implementation of span copied over from array_view. It
+     * returns non-const iterator/pointers. first(), last(), and subspan() 
+     * are yet to be implemented. It does not maintain storage for the data, 
+     * but just holds pointers to mark the beginning and end of the array. 
+     * It can be conveniently constructed from a variety of other container 
+     * types like array, vector, and fixed_vector. 
+     *
+     * Example:
+     *    Given "void Func(AZStd::span<int> a) {...}" you can call...
+     *    - Func({1,2,3});
+     *    - AZStd::array<int,3> a = {1,2,3};
+     *      Func(a);
+     *    - AZStd::vector<int> v = {1,2,3};
+     *      Func(v);
+     *    - AZStd::fixed_vector<int,10> fv = {1,2,3};
+     *      Func(fv);
+     *
+     * Since the span does not copy and store any data, it is only valid as long as the data used to create it is valid.
+     */
+    template <class Element>
+    class span final
+    {
+    public:
+        using value_type = Element;
+
+        using pointer = value_type*;
+        using const_pointer = const value_type*;
+
+        using reference = value_type&;
+        using const_reference = const value_type&;
+
+        using size_type = AZStd::size_t;
+        using difference_type = AZStd::ptrdiff_t;
+
+        using iterator = value_type*;
+        using const_iterator = const value_type*;
+        using reverse_iterator = AZStd::reverse_iterator<iterator>;
+        using const_reverse_iterator = AZStd::reverse_iterator<const_iterator>;
+
+        span()
+            : m_begin(nullptr)
+            , m_end(nullptr)
+        { }
+
+        ~span() = default;
+
+        span(const_pointer s, size_type length)
+            : m_begin(s)
+            , m_end(m_begin + length)
+        {
+            if (length == 0) erase();
+        }
+
+        span(const_pointer first, const_pointer last)
+            : m_begin(first)
+            , m_end(last)
+        { }
+
+        // We explicitly delete this constructor because it's too easy to accidentally 
+        // create a span to just the first element instead of an entire array.
+        span(const_pointer s) = delete;
+
+        template<AZStd::size_t N>
+        span(AZStd::array<value_type, N>& data)
+            : m_begin(data.data())
+            , m_end(m_begin + data.size())
+        { }
+
+        span(AZStd::vector<value_type>& data)
+            : m_begin(data.data())
+            , m_end(m_begin + data.size())
+        { }
+
+        template<AZStd::size_t N>
+        span(AZStd::fixed_vector<value_type, N>& data)
+            : m_begin(data.data())
+            , m_end(m_begin + data.size())
+        { }
```

Comment:
I think you also want versions of these that take in const references, for the cases that you're trying to construct a span with const elements (like what array_view currently does).  FWIW, you could do it like this:
```
        template<typename Container>
        array_view(Container& data)
            : m_begin(data.data())
            , m_end(m_begin + data.size())
        { }

        template<typename Container>
        array_view(const Container& data)
            : m_begin(data.data())
            , m_end(m_begin + data.size())
        { }
```

Reply:
Added.